knowledge_repo/app/app.py

```diff
@@ -30,26 +36,35 @@ def __init__(self, repo, db_uri=None, debug=None, config=None, **kwargs):
                        template_folder='templates',
                        static_folder='static')
 
-        # Load configuration
+        # Preload default configuration
+        self.config.from_object('knowledge_repo.app.config_defaults')
+
+        # Load configuration from file or provided object
         if config:
             if isinstance(config, str):
                 config = imp.load_source('knowledge_server_config', os.path.abspath(config))
             self.config.from_object(config)
+
+        # Add configuration passed in as keyword arguments
         self.config.update(kwargs)
 
+        # Prepare repository, and add it to the app
         if hasattr(config, 'prepare_repo'):
             repo = config.prepare_repo(repo)
         self.repository = repo
+        assert isinstance(self.repository, knowledge_repo.KnowledgeRepository), "Invalid repository object provided."
 
-        # Lift debug mode from underlying KnowledgeRepository object if not specified
-        self.config['DEBUG'] = debug if debug is not None else repo.config.debug
+        # Set debug mode from kwargs or else maintain current setting
+        if debug is not None:
+            self.config['DEBUG'] = debug
+
+        # Set the secret key for this instance (creating one if one does not exist already)
+        self.config['SECRET_KEY'] = self.config['SECRET_KEY'] or str(uuid.uuid4())
```

Comment:
The secret key is used by flask-login to encrypt cookies so that it's not possible to modify authentication details.